### PR TITLE
feat: support for retrieving Advanced search definition in REST Module

### PR DESCRIPTION
**Pull request by PenghaiZhang**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [ ] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This PR adds the support for retrieving Advanced search to the REST Module. 

The major change is creating different data structures to support Wizard controls used in Advanced search. And a new function was added to talk to the endpoint.

### CODE REVIEW

File: oeq-ts-rest-api/src/AdvancedSearch.ts [hunk 1]
```diff
@@ -15,8 +15,32 @@
...
  */
+import { is } from 'typescript-is';
 import { GET } from './AxiosInstance';
 import { BaseEntitySummary, isBaseEntitySummaryArray } from './Common';
+import { WizardControl } from './wizard/WizardControl';
+
+/**
+ * Definition of an Advanced Search.
+ */
+export interface AdvancedSearch {
```
Comment (PenghaiZhang) [inline oeq-ts-rest-api/src/AdvancedSearch.ts:26]: Probably need a different name for this type...

I am working on another branch created off this branch and I have this awkward naming.
```
Promise<OEQ.AdvancedSearch.AdvancedSearch>
```

Comment (PenghaiZhang) [reply oeq-ts-rest-api/src/AdvancedSearch.ts:26]: The namespace is very clear to say this is `AdvancedSearch`  so maybe  rename the type to  `Definition`?

Comment (edalex-ian) [reply oeq-ts-rest-api/src/AdvancedSearch.ts]: Maybe, I'm not to worried either way.

Comment (PenghaiZhang) [reply oeq-ts-rest-api/src/AdvancedSearch.ts]: I renamed it to `WizardDefinition`.

File: oeq-ts-rest-api/src/AdvancedSearch.ts [hunk 2]
```diff
@@ -15,8 +15,32 @@
...
  */
+import { is } from 'typescript-is';
 import { GET } from './AxiosInstance';
 import { BaseEntitySummary, isBaseEntitySummaryArray } from './Common';
+import { WizardControl } from './wizard/WizardControl';
+
+/**
+ * Definition of an Advanced Search.
+ */
+export interface WizardDefinition {
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/AdvancedSearch.ts:26]: Nope, can't use this name, this is _not_ a definition of a wizard, it is the definition of an Advanced Search.

Here's some options in order of preference:

- `AdvancedSearchDefinition`
- `AdvancedSearch` - what you originally had, I'm not too concerned about the repitition
- `Definition`

File: oeq-ts-rest-api/src/AdvancedSearch.ts [hunk 3]
```diff
@@ -15,8 +15,32 @@
...
+  /**
+   * Name of an Advanced Search.
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/AdvancedSearch.ts:28]:
```suggestion
   * Name of the Advanced Search.
```

File: oeq-ts-rest-api/src/AdvancedSearch.ts [hunk 4]
```diff
@@ -15,8 +15,32 @@
...
+   */
+  name?: string;
+  /**
+   * Description of an Advanced Search.
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/AdvancedSearch.ts:32]:
```suggestion
   * Description of the Advanced Search.
```

File: oeq-ts-rest-api/src/AdvancedSearch.ts [hunk 5]
```diff
@@ -15,8 +15,32 @@
...
+   */
+  description?: string;
+  /**
+   * A list of Collections configured to limit the search result.
+   */
+  collections: BaseEntitySummary[];
+  /**
+   * All the Wizard Controls of an Advanced Search.
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/AdvancedSearch.ts:40]:
```suggestion
   * All the Wizard Controls of the Advanced Search.
```

File: oeq-ts-rest-api/src/Taxonomy.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Restrictions applied to Taxonomy term selection.
+ */
+export enum SelectionRestriction {
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/Taxonomy.ts:22]: A reminder that we're avoiding the use of `enum`s on the TS side - due to poor support and possible future clashes with ESCMA versions.

Typically instead we've gone with a simple object definition. e.g.

```typescript
export const SelectionRestriction = {
  TOP_LEVEL_ONLY: 'TOP_LEVEL_ONLY',
  LEAF_ONLY: 'LEAF_ONLY',
  UNRESTRICTED: 'UNRESTRICTED'
}
```

If you wish, you could also add typing to then mark them as `readonly` but I don't think we bothered elsewhere.

File: oeq-ts-rest-api/src/Taxonomy.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  TOP_LEVEL_ONLY = 'TOP_LEVEL_ONLY',
+  LEAF_ONLY = 'LEAF_ONLY',
+  UNRESTRICTED = 'UNRESTRICTED',
+}
+
+/**
+ * Formats which ae used to search for a term.
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/Taxonomy.ts:29]:
```suggestion
 * Formats which are used to search for a term.
```

File: oeq-ts-rest-api/src/Taxonomy.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ */
+export enum TermStorageFormat {
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/Taxonomy.ts:31]: As above: A reminder that we're avoiding the use of `enum`s on the TS side - due to poor support and possible future clashes with ESCMA versions.

File: oeq-ts-rest-api/src/wizard/WizardCommonTypes.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+ */
+export interface TargetNode {
+  /**
+   * The target node.
+   */
+  target: string;
+  /**
+   * Attributes of the target node.
+   */
+  attribute: string;
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/wizard/WizardCommonTypes.ts:30]: Odd that the above comment uses plural but this is both singular in name and type. :thinking:

File: oeq-ts-rest-api/src/wizard/WizardCommonTypes.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  /**
+   * Full path of the target node.
+   */
+  fullTarget: string;
+  /**
+   * XOQ path of the target node.
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/wizard/WizardCommonTypes.ts:36]:
```suggestion
   * XOQL path of the target node.
```

File: oeq-ts-rest-api/src/wizard/WizardCommonTypes.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+   */
+  xoqlPath: string;
+  /**
+   * Free text field of the target node.
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/wizard/WizardCommonTypes.ts:40]: if I follow this right, perhaps this should be:

```suggestion
   * Free text search index field of the target node.
```

File: oeq-ts-rest-api/src/wizard/WizardCommonTypes.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+   */
+  freetextField: string;
+}
+
+/**
+ * Options of Option' type control such as CheckBox Group and Shuffle List.
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/wizard/WizardCommonTypes.ts:46]:
```suggestion
 * Options of 'Option' type control such as CheckBox Group and Shuffle List.
```

File: oeq-ts-rest-api/src/wizard/WizardControl.ts [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { UuidString } from '../Common';
+import { SelectionRestriction, TermStorageFormat } from '../Taxonomy';
+import { TargetNode, WizardControlOption } from './WizardCommonTypes';
+
+export type ControlType =
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/wizard/WizardControl.ts:22]: Needs doc. Maybe:

```suggestion
/**
 * Supported Wizard Control types.
 */
export type ControlType =
```

File: oeq-ts-rest-api/src/wizard/WizardControl.ts [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+
+export interface WizardUserSelectorControl extends WizardBasicControl {
+  /**
+   * Attributes specific to UserSelector.
+   */
+  attributes: {
+    /**
+     * Whether to restrict user selection by groups
+     */
+    IsRestrictedGroups: boolean;
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/wizard/WizardControl.ts:146]:
```suggestion
    isRestrictedGroups: boolean;
```

File: oeq-ts-rest-api/src/wizard/WizardControl.ts [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+    /**
+     * Whether selecting multiple users is supported.
+     */
+    IsSelectMultiple: boolean;
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/wizard/WizardControl.ts:150]:
```suggestion
    isSelectMultiple: boolean;
```

File: oeq-ts-rest-api/src/wizard/WizardControl.ts [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+    /**
+     * Groups which the selection is limited to.
+     */
+    RestrictToGroups: UuidString[];
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/wizard/WizardControl.ts:154]:
```suggestion
    restrictToGroups: UuidString[];
```

File: oeq-ts-rest-api/src/wizard/WizardControl.ts [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+  };
+}
+
+export interface WizardTermSelectorControl extends WizardBasicControl {
+  /**
+   * Attributes specific to TermSelector.
+   */
+  attributes: {
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/wizard/WizardControl.ts:162]: The names and casing of the properties of this are rather odd. What's going on here? :thinking:

File: oeq-ts-rest-api/src/wizard/WizardControl.ts [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+   */
+  options: WizardControlOption[];
+  /**
+   * Text displayed in the criteria summary instead.
+   */
+  powerSearchFriendlyName?: string;
+  /**
+   * Type of the control.
+   */
+  controlType: ControlType;
```
Comment (edalex-ian) [inline oeq-ts-rest-api/src/wizard/WizardControl.ts:89]: It's a little interesting that you didn't set the explicit value for this in each interface below - which could then be used as a union discriminator. Why was that?

**Review by edalex-ian — CHANGES_REQUESTED** [review]
Thanks @PenghaiZhang .

Most of the feedback is to tweak comments. But other than that:

- Some queries around some of the data definitions; and
- Request to continue to avoid TS enums.

File: oeq-ts-rest-api/src/wizard/WizardControl.ts (hunk 17 above)
Comment (PenghaiZhang) [reply oeq-ts-rest-api/src/wizard/WizardControl.ts:162]: I have refactored both server and client for this type :smile:

File: oeq-ts-rest-api/src/Taxonomy.ts (hunk 6 above)
Comment (PenghaiZhang) [reply oeq-ts-rest-api/src/Taxonomy.ts:22]: Can we just create a type instead ? 

```
export type TermStorageFormat = 'FULL_PATH' | 'LEAF_ONLY';
```

File: oeq-ts-rest-api/src/wizard/WizardCommonTypes.ts (hunk 9 above)
Comment (PenghaiZhang) [reply oeq-ts-rest-api/src/wizard/WizardCommonTypes.ts:30]: I think this field is a singular so I will update the JSDoc.

File: oeq-ts-rest-api/src/wizard/WizardControl.ts (hunk 18 above)
Comment (PenghaiZhang) [reply oeq-ts-rest-api/src/wizard/WizardControl.ts:89]: I was thinking to create some type guards in future work.

Comment (PenghaiZhang) [reply oeq-ts-rest-api/src/wizard/WizardControl.ts:89]: Not a problem. I will add explicit control type in each type definition.

**Comment by PenghaiZhang** [comment]
I have updated the data structure for `TermSelector` and `UserSelector`. 
Screenshot from 2021-09-15 10-51-22

File: oeq-ts-rest-api/src/Taxonomy.ts (hunk 6 above)
Comment (edalex-ian) [reply oeq-ts-rest-api/src/Taxonomy.ts]: Actually, yes. That might be the approach we've been using - I was hoping to find you an example, but I can't remember... :confused:

**Review by edalex-ian — APPROVED** [review]
Looks great @PenghaiZhang ! Merge on! :+1:

**Pull request merged**